src/presentation-2/upgrader.ts

```diff
@@ -714,3 +762,40 @@ export function convertPresentation2(entity: any): Presentation3.Manifest | Pres
   }
   return entity;
 }
+
+function upgradeSelector(
+  selector: Presentation2.ContentResourceSelector
+): Presentation3.Selector | Presentation3.Selector[] {
+  const isSvgSelector =
+    ((Array.isArray(selector['@type']) && selector['@type'].includes('oa:SvgSelector')) ||
+      selector['@type'] == 'oa:SvgSelector') &&
+    ('chars' in selector || 'value' in selector);
+  if (isSvgSelector) {
+    return {
+      type: 'SvgSelector',
+      value: 'chars' in selector ? selector.chars : selector.value,
+    };
+  }
+  if (selector['@type'] === 'oa:FragmentSelector') {
+    return {
+      type: 'FragmentSelector',
+      value: selector.value,
+    };
+  }
+  if (selector['@type'] === 'oa:Choice') {
+    return [
+      upgradeSelector(selector.default) as Presentation3.Selector,
+      ...((Array.isArray(selector.item) ? selector.item : [selector.item]).map(
+        upgradeSelector
+      ) as Presentation3.Selector[]),
+    ];
+  }
+  if (selector['@type'] == 'iiif:ImageApiSelector') {
+    return {
+      type: 'ImageApiSelector',
+      region: 'region' in selector ? selector.region : undefined,
+      rotation: 'rotation' in selector ? selector.rotation : undefined,
+    };
```

Comment:
Am I correct in thinking that this is based on the cookbook: https://github.com/IIIF/trc/issues/97 ?

Reply:
I was basing it on the existing types in `@iiif/presentation-2-types` (https://github.com/IIIF-Commons/presentation-2-types/blob/main/resources/content-resource.d.ts#L7-L26) and cross-checking those with the specification at https://iiif.io/api/annex/openannotation/.
The specification has a bunch more fields, should we maybe add those to the `presentation-2-types` and then upgrade them here as well? The spec of the selector pre-dates IIIF3 (v1.0 was 2014).

Reply:
Hmm, difficult one. The cookbook doesn't use the additional fields - and I don't know how common they will end up being (especially for 2->3 conversion).

The cookbook includes the context inside the selector and keeps the prefixed `iiif:ImageApiSelector`
```json
{
    "@context": "http://iiif.io/api/annex/openannotation/context.json",
    "type": "iiif:ImageApiSelector",
    ...
}
```